File: 1242-web-crawler-multithreaded/1242-web-crawler-multithreaded.java
/**
 * // This is the HtmlParser's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface HtmlParser {
 *     public List<String> getUrls(String url) {}
 * }
 */
class Solution {
    private int maxNumberOrWorkers = 8;
    
    private void work(int workerId, String hostName, Queue<String> urlQ, HashSet<String> visUrl, ReentrantLock urlQLock, Condition urlQCdn, AtomicInteger idleWorkers, HtmlParser htmlParser) {
        while (true) {
            String currUrl = null;
            
            urlQLock.lock();
            try {
                while (urlQ.isEmpty()) {
                    // starting to be idle
                    idleWorkers.getAndIncrement();
                    if (idleWorkers.get() == maxNumberOrWorkers) {
                        urlQCdn.signalAll();
                        return;
                    }
                    
                    urlQCdn.await();
                    
                    // might become non idle
                    idleWorkers.getAndDecrement();
                }
                
                currUrl = urlQ.poll();
            } catch (InterruptedException e) {
                e.printStackTrace();  
            } finally {
                urlQLock.unlock();
            }
            
            if (currUrl != null) {
                List<String> nextUrls = htmlParser.getUrls(currUrl);
                
                urlQLock.lock();
                try {
                    for (String nextUrl: nextUrls) {
                        String localHostName = getHostName(nextUrl);
                        
                        if (!visUrl.contains(nextUrl) && localHostName.equals(hostName)) {
                            urlQ.add(nextUrl);
                            visUrl.add(nextUrl);
                            
                            urlQCdn.signal();
                        }
                    }
                } finally {
                    urlQLock.unlock();
                }
            }
        }
    }
    
    private String getHostName(String url) {
        String hostName = url.substring(7).split("/")[0];
        return hostName;
    }
    
    public List<String> crawl(String startUrl, HtmlParser htmlParser) {
        
        String hostName = getHostName(startUrl);
        
        Queue<String> urlQ = new LinkedList<>();
        HashSet<String> visUrl = new HashSet<>();
        
        urlQ.offer(startUrl);
        visUrl.add(startUrl);
        
        Thread[] workers = new Thread[maxNumberOrWorkers];
        ReentrantLock urlQLock = new ReentrantLock();
        Condition urlQCdn = urlQLock.newCondition();
        AtomicInteger idleWorkers = new AtomicInteger(0);
        
        for (int i = 0; i < maxNumberOrWorkers; i++) {
            int workerId = i;
            
            Thread worker = new Thread(() -> work(workerId, hostName, urlQ, visUrl, urlQLock, urlQCdn, idleWorkers, htmlParser));
            workers[workerId] = worker;
            workers[workerId].start();
        }
        
        for (int i = 0; i < maxNumberOrWorkers; i++) {
            int workerId = i;
            
            try {
                workers[workerId].join();    
            } catch (InterruptedException e) {
                e.printStackTrace();  
            }
        }
        
        return new ArrayList<>(visUrl);
    }
}

File: 1242-web-crawler-multithreaded/1242-web-crawler-multithreaded.cpp
class Solution {
private:
    const int maxWorker = 5;
    
    string getHostname(const string& url) {
        int startIdx = url.find("//") + 2;
        int endIdx = url.find('/', startIdx);

        if (endIdx == string::npos) {
            return url.substr(startIdx);
        }

        return url.substr(startIdx, endIdx - startIdx);
    }

public:
    vector<string> crawl(string startUrl, HtmlParser htmlParser) {
        string hostName = getHostname(startUrl);
        unordered_set<string> parsed;
        queue<string> urlQ;
        
        urlQ.push(startUrl);
        parsed.insert(startUrl);
        
        mutex urlQLock;
        condition_variable urlQCdn;
        set<int> idleWorker;
        
        vector<thread> workers;
        
        for (int i = 0; i < maxWorker; ++i) {
            workers.emplace_back([&, i]() {
                while (true) {
                    string currUrl;
                    {
                        unique_lock<mutex> lock(urlQLock);
                        urlQCdn.wait(lock, [&] {
                            if (urlQ.empty()) {
                                idleWorker.insert(i);
                                if (idleWorker.size() == maxWorker) {
                                    urlQCdn.notify_all();
                                    return true;
                                }
                            }
                            return !urlQ.empty();
                        });

                        if (idleWorker.size() == maxWorker) {
                            return;
                        }

                        currUrl = urlQ.front();
                        urlQ.pop();
                    }

                    // get next urls
                    vector<string> nextUrls = htmlParser.getUrls(currUrl);

                    unique_lock<mutex> lock(urlQLock);
                    for (const string& nurl : nextUrls) {
                        string localHostName = getHostname(nurl);
                        if (parsed.find(nurl) == parsed.end() && localHostName == hostName) {
                            urlQ.push(nurl);
                            parsed.insert(nurl);
                            urlQCdn.notify_one();
                        }
                    }
                    
                    if (urlQ.empty()) {
                        idleWorker.insert(i);
                        urlQCdn.notify_all();
                    } else {
                        idleWorker.erase(i);
                    }
                }
            });
        }
        
        for (auto& worker : workers) {
            worker.join();
        }
        
        return vector<string>(parsed.begin(), parsed.end());
    }
};